File: hello-python/src/audio_server/audio_server_application.py
#!/usr/bin/env python3
"""
Audio Stream Server - Python Implementation
Main entry point for the server
"""

import asyncio
import argparse
from loguru import logger
import sys

from .network.audio_websocket_server import AudioWebSocketServer
from .memory.stream_manager import StreamManager
from .memory.memory_pool_manager import MemoryPoolManager


def parse_args():
    """Parse command line arguments"""
    parser = argparse.ArgumentParser(description="Audio Stream Server - Python Implementation")
    parser.add_argument("--host", default="0.0.0.0", help="Host address to bind to (default: 0.0.0.0)")
    parser.add_argument("--port", type=int, default=8080, help="Port number to listen on (default: 8080)")
    parser.add_argument("--path", default="/audio", help="WebSocket endpoint path (default: /audio)")
    parser.add_argument("--cache-dir", default="cache", help="Cache directory for stream files (default: cache)")
    parser.add_argument("--buffer-size", type=int, default=65536, help="Buffer size in bytes (default: 64KB)")
    parser.add_argument("--pool-size", type=int, default=100, help="Number of buffers in pool (default: 100)")
    parser.add_argument("--verbose", "-v", action="store_true", help="Enable verbose logging")
    return parser.parse_args()


async def main():
    """Main entry point"""
    args = parse_args()
    
    # Configure logging
    logger.remove()  # Remove default handler
    if args.verbose:
        logger.add(sys.stderr, level="DEBUG")
    else:
        logger.add(sys.stderr, level="INFO")
    
    logger.info("=" * 60)
    logger.info("Audio Stream Server - Python Implementation")
    logger.info("=" * 60)
    
    # Initialize singletons
    logger.info("Initializing server components...")
    stream_manager = StreamManager.get_instance(cache_directory=args.cache_dir)
    memory_pool = MemoryPoolManager.get_instance(buffer_size=args.buffer_size, pool_size=args.pool_size)
    
    logger.info(f"Stream Manager: cache directory = {args.cache_dir}")
    logger.info(f"Memory Pool: {args.pool_size} buffers × {args.buffer_size} bytes")
    
    # Create and start WebSocket server
    server = AudioWebSocketServer(host=args.host, port=args.port, path=args.path)
    await server.start()
    
    logger.info("=" * 60)
    logger.info(f"Server ready at ws://{args.host}:{args.port}{args.path}")
    logger.info("Press Ctrl+C to stop")
    logger.info("=" * 60)
    
    # Keep server running
    try:
        await asyncio.Future()  # Run forever
    except KeyboardInterrupt:
        logger.info("\nShutting down server...")
        await server.stop()
        logger.info("Server stopped")


if __name__ == "__main__":
    asyncio.run(main())

File: hello-rust/src/server/audio_server_application.rs
// Main entry point for Rust audio stream server.

use crate::server::memory::{MemoryPoolManager, StreamManager};
use crate::server::network::AudioWebSocketServer;
use std::env;

#[allow(dead_code)]
pub fn run_server() {
    // Parse command-line arguments
    let args: Vec<String> = env::args().collect();
    let mut port = 8080;
    let mut path = "/audio".to_string();

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--port" => {
                if i + 1 < args.len() {
                    port = args[i + 1].parse().unwrap_or(8080);
                    i += 1;
                }
            }
            "--path" => {
                if i + 1 < args.len() {
                    path = args[i + 1].clone();
                    i += 1;
                }
            }
            _ => {}
        }
        i += 1;
    }

    println!("Starting Audio Server on port {} with path {}", port, path);

    // Get singleton instances
    let stream_manager = StreamManager::instance("cache".to_string());
    let memory_pool = MemoryPoolManager::instance(65536, 100);

    // Create and start WebSocket server
    let ws_server = AudioWebSocketServer::new(port, path, stream_manager, memory_pool);

    ws_server.start();
}
